### PR TITLE
[chore] add pre commit

### LINKED ISSUE #194: Add pre-commit
put black, flake8, isort, and clang as pre-commit

### PR DESCRIPTION
**Pull request by abhinavDhulipala**
Couple changes(improvements) to the dev experience here. Workflow compute isn't wasted if code-checks (static lints) don't pass. pre-commit is automatically run upon every commit (given `pre-commit install`). It's extremely fast. On my 2vCPU code space it runs in less than a second for **all files** (`pre-commit run -a`). This is the very upper-bound as developers will only have pre-commit run on staged files. I also believe it is slightly more efficient than the existing docker-compose based linting solution. I checked that every one of the hooks I used are officially supported and/or part of active, healthy repos.

closes #194

### CODE REVIEW

**Comment by abhinavDhulipala** [comment]
Woops need to rebase, srry

**Comment by abhinavDhulipala** [comment]
Here's the context for the isort [cfg](https://github.com/PyCQA/isort/issues/1889)

**Comment by abhinavDhulipala** [comment]
@WillAyd lmk if the changes to the [CONTRIBUTING.md](https://github.com/innobi/pantab/pull/195/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055) are sufficient/clear enough. Also if you could double check the clang pre-commit that'd be great. i.e pull the branch, violate clang-15, and ensure it fails as you expect. Not really an expert in that area. Just copied the configs you had previously on your lint [stage](https://github.com/innobi/pantab/blob/main/ci/scripts/lint_check.sh#L7)

File: .github/workflows/pythonpackage.yml [hunk 1]
```diff
@@ -10,20 +10,17 @@ on:
...
-
     steps:
     - uses: actions/checkout@v2
-    - name: Run lint checks
-      run: docker compose run --rm lint ci/scripts/lint_check.sh
+    - uses: pre-commit/action@v3.0.0
 
   build-linux:
-
+    needs: code-checks
```
Comment (WillAyd) [inline .github/workflows/pythonpackage.yml:18]: we arguably don't need to do this every run - maybe worth its own job?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+repos:
```
Comment (WillAyd) [inline .pre-commit-config.yaml:1]: Its probably worth trying to sync with pandas on this, which uses one repo for most checks

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  - repo: https://github.com/ambv/black
+    rev: 23.9.1
+    hooks:
+    - id: black
+  - repo: https://github.com/pycqa/isort
+    rev: 5.12.0
+    hooks:
+      - id: isort
+  - repo: https://github.com/pycqa/flake8
```
Comment (WillAyd) [inline .pre-commit-config.yaml:10]: We could also replace this with ruff

**Comment by WillAyd** [comment]
@abhinavDhulipala this is great - thanks for the PR! Let's give this a shot. We can try some improvements in follow ups but this alone should really help modernize our stack as is

**Pull request merged**